src/pingttl.nim

```diff
@@ -0,0 +1,270 @@
+    if msg.msg_controllen == 0:
+      trace("pingttl: errqueue control msg is empty")
+      raiseOSError(lastError)
+    var cmsg = CMSG_FIRSTHDR(addr(msg))
+    while cmsg != nil:
+      # https://www.man7.org/linux/man-pages/man7/ip.7.html
+      # > Enable extended reliable error message passing.  When
+      # > enabled on a datagram socket, all generated errors will be
+      # > queued in a per-socket error queue.  When the user
+      # > receives an error from a socket operation, the errors can
+      # > be received by calling recvmsg(2) with the MSG_ERRQUEUE
+      # > flag set.  The sock_extended_err structure describing the
+      # > error will be passed in an ancillary message with the type
+      # > IP_RECVERR and the level IPPROTO_IP.  This is useful for
+      # > reliable error handling on unconnected sockets.  The
+      # > received data portion of the error queue contains the
+      # > error packet.
+      if cmsg.cmsg_len   > 0 and
+         cmsg.cmsg_level == IPPROTO_IP and
+         cmsg.cmsg_type  == IP_RECVERR:
+        let err = cast[ptr SockExtendedErr](CMSG_DATA(cmsg))
+        if err.ee_origin != SO_EE_ORIGIN_ICMP:
+          # this can be either icmp6 or local errors
+          trace("pingttl: control message error origin is not icmp error " & $err.ee_origin)
+          raiseOSError(lastError)
+        if (err.ee_type != uint8(IcmpType.Unreachable) and
+            err.ee_type != uint8(IcmpType.TTLExceeded)):
+          trace("pingttl: control message icmp type is neither unreachable or ttl exceeded " & $err.ee_type)
+          raiseOSError(lastError)
+        # https://www.man7.org/linux/man-pages/man7/ip.7.html
+        # > ee_errno contains the errno number of the queued error.
+        # > ee_origin is the origin code of where the error
+        # > originated.  The other fields are protocol-specific.  The
+        # > macro SO_EE_OFFENDER returns a pointer to the address of
+        # > the network object where the error originated from given a
+        # > pointer to the ancillary message.  If this address is not
+        # > known, the sa_family member of the sockaddr contains
+        # > AF_UNSPEC and the other fields of the sockaddr are
+        # > undefined.
+        let offender = SO_EE_OFFENDER(err)
+        if int32(offender.sa_family) == posix.AF_UNSPEC:
+          trace("pingttl: icmp error from unknown offender source address family " & $offender.sa_family)
+          raiseOSError(lastError)
+        if int32(offender.sa_family) != posix.AF_INET:
+          trace("pingttl: icmp error from unsupported offender source address family " & $offender.sa_family)
+          raiseOSError(lastError)
+        # TODO how to get the sock len directly from the sockaddr structure?
+        # var
+        #   address: IpAddress
+        #   port:    Port
+        # fromSockAddr(cast[ptr Sockaddr_in](offender)[], ???, address, port)
+        let address = cast[ptr Sockaddr_in](offender)
```

Comment:
My recollection is that Nim has a sizeof operator like C, so you can just alloc something of the right size. You also can just always allocate to the size of the `sockaddr_in6` struct, and pass it; the over-allocation isn't an issue when it's an ipv4 connection (in fact, that's essentially how I do it in libcon4m for the IP address type, I always allocate `sizeof(struct sockaddr_in6)`).  But this is all easy to do from C too, if needed.

Reply:
The big Nim gotcha BTW is that you have to worry about the Nim lifetime for anything that's getting passed to C. Probably won't be an issue here, but there are odd scenarios where the Nim semantics won't be obvious.

Reply:
thanks. directly passing sizeof ipv4 worked. didnt go for the ipv6 as above error handling already asserts its ipv4 error as otherwise well need to handle ipv6 errors differently. did add a comment so when we need to add ipv6 support it shouldnt be too bad. just diff error codes and sock options